storage/integrate.go

```diff
@@ -0,0 +1,337 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to create fulltile: %v", err)
+		}
+		readCache.Set(tileID, treeSize, ft)
+		return ft, nil
+	}
+	return &TreeBuilder{
+		getTile: getFullTile,
+		rf:      &compact.RangeFactory{Hash: rfc6962.DefaultHasher.HashChildren},
+	}
+}
+
+// newRange creates a new compact.Range for the specified treeSize, fetching tiles as necessary.
+func (t *TreeBuilder) newRange(ctx context.Context, treeSize uint64) (*compact.Range, error) {
+	rangeNodes := compact.RangeNodes(0, treeSize, nil)
+	errG := errgroup.Group{}
+	hashes := make([][]byte, len(rangeNodes))
+	for i, id := range rangeNodes {
+		i := i
+		id := id
+		errG.Go(func() error {
+			tLevel, tIndex, nLevel, nIndex := layout.NodeCoordsToTileAddress(uint64(id.Level), id.Index)
+			ft, err := t.getTile(ctx, TileID{Level: tLevel, Index: tIndex}, treeSize)
+			if err != nil {
+				return err
+			}
+			h := ft.Get(compact.NodeID{Level: nLevel, Index: nIndex})
+			if h == nil {
+				return fmt.Errorf("missing node: [%d/%d@%d]", id.Level, id.Index, treeSize)
+			}
+			hashes[i] = h
+			return nil
+		})
+	}
+	if err := errG.Wait(); err != nil {
+		return nil, err
+	}
+	return t.rf.NewRange(0, treeSize, hashes)
+}
+
+func (t *TreeBuilder) Integrate(ctx context.Context, fromSize uint64, entries [][]byte) (newSize uint64, rootHash []byte, tiles map[TileID]*api.HashTile, err error) {
+	baseRange, err := t.newRange(ctx, fromSize)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to create range covering existing log: %w", err)
+	}
+
+	// Initialise a compact range representation, and verify the stored state.
+	r, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("invalid log state, unable to recalculate root: %w", err)
+	}
+
+	klog.V(1).Infof("Loaded state with roothash %x", r)
+	// Create a new compact range which represents the update to the tree
+	newRange := t.rf.NewEmptyRange(fromSize)
+	if len(entries) == 0 {
+		klog.V(1).Infof("Nothing to do.")
+		// Nothing to do, nothing done.
+		return fromSize, r, nil, nil
+	}
+	tc := newTileWriteCache(fromSize, t.getTile)
+	visitor := tc.Visitor(ctx)
+	for _, e := range entries {
+		lh := rfc6962.DefaultHasher.HashLeaf(e)
+		// Update range and set nodes
+		if err := newRange.Append(lh, visitor); err != nil {
+			return 0, nil, nil, fmt.Errorf("newRange.Append(): %v", err)
+		}
+
+	}
+
+	// Merge the update range into the old tree
+	if err := baseRange.AppendRange(newRange, visitor); err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to merge new range onto existing log: %w", err)
+	}
+
+	// Calculate the new root hash - don't pass in the tileCache visitor here since
+	// this will construct any ephemeral nodes and we do not want to store those.
+	newRoot, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to calculate new root hash: %w", err)
+	}
+
+	if err := tc.Err(); err != nil {
+		return 0, nil, nil, err
+	}
```

Comment:
Worth moving this up a few lines? Maybe even checking it before and after `baseRange.AppendRange(newRange, visitor)` ? Reasoning being that any errors in here are likely to cause failures in the other operations that depend on the writeTileCache succeeding. It's better to return the root error rather than a symptomatic error.

Reply:
done